bounded/src/set.rs

```diff
@@ -511,332 +608,267 @@ fn nullary_id_to_app(term: Term, relations: &[RelationDecl]) -> Term {
-        Term::NAryOp(NOp::Or, terms) => or(terms.iter().map(ast).collect::<Result<_, _>>()?),
-        Term::Ite { cond, then, else_ } => or(btreeset![
-            and(btreeset![ast(cond)?, ast(then)?]),
-            and(btreeset![not(ast(cond)?)?, ast(else_)?])
-        ]),
+        Term::NAryOp(NOp::Or, terms) => {
+            Formula::or(terms.iter().map(formula).collect::<Result<Vec<_>, _>>()?)
+        }
+        Term::Ite { cond, then, else_ } => formula(cond)?.ite(formula(then)?, formula(else_)?),
         Term::Quantified {
             quantifier,
             binders,
             body,
         } => {
-            // adapted from semantics.rs
-            assert!(!binders.is_empty());
-            let shape: Vec<usize> = binders
+            let terms = binders
                 .iter()
-                .map(|b| cardinality(universe, &b.sort))
-                .collect();
-            let names: Vec<&String> = binders.iter().map(|b| &b.name).collect();
-            // evaluate on all combinations of values for quantified sorts
-            let set = shape
-                .iter()
-                .map(|&card| (0..card).collect::<Vec<Element>>())
+                .map(|b| cardinality(context.universe, &b.sort))
+                .map(|card| (0..card).collect::<Vec<Element>>())
                 .multi_cartesian_product_fixed()
                 .map(|elements| {
-                    // extend context with all variables bound to these `elements`
                     let mut new_assignments = assignments.clone();
-                    for (name, element) in names.iter().zip(elements) {
-                        new_assignments.insert(name.to_string(), element);
+                    assert_eq!(binders.len(), elements.len());
+                    for (binder, element) in binders.iter().zip(elements) {
+                        new_assignments.insert(binder.name.to_string(), element);
                     }
-                    term_to_ast(body, universe, &new_assignments)
+                    term_to_formula(body, context, &new_assignments)
                 })
-                .collect::<Result<BTreeSet<_>, _>>()?;
+                .collect::<Result<Vec<_>, _>>()?;
             match quantifier {
-                Quantifier::Forall => and(set),
-                Quantifier::Exists => or(set),
+                Quantifier::Forall => Formula::and(terms),
+                Quantifier::Exists => Formula::or(terms),
             }
         }
         Term::UnaryOp(UOp::Prime | UOp::Always | UOp::Eventually, _)
         | Term::UnaryOp(UOp::Next | UOp::Previous, _)
         | Term::BinOp(BinOp::Until | BinOp::Since, ..)
-        | Term::App(..) => return Err(TranslationError::CouldNotTranslateToAst(term.clone())),
+        | Term::App(..) => return Err(TranslationError::CouldNotTranslateToFormula(term.clone())),
     };
-    Ok(ast)
+    Ok(formula)
 }
 
 fn term_to_element(
```

Comment:
Is it correct to say that this function tries to evaluate the given term to a constant (which might be a bool or an uninterpreted sort)? Can you document this?